src/AsyncProducers.cpp

```diff
@@ -0,0 +1,598 @@
+    using IRMutator::visit;
+
+    const map<string, Function> &env;
+
+    map<string, string> cloned_acquires;
+
+    void visit(const Realize *op) {
+        auto it = env.find(op->name);
+        internal_assert(it != env.end());
+        Function f = it->second;
+        if (f.schedule().async()) {
+            Stmt body = op->body;
+
+            // Make two copies of the body, one which only does the
+            // producer, and one which only does the consumer. Inject
+            // synchronization to preserve dependencies. Put them in a
+            // task-parallel block.
+
+            // Make a semaphore per consume node
+            CountConsumeNodes consumes(op->name);
+            body.accept(&consumes);
+
+            vector<string> sema_names;
+            vector<Expr> sema_vars;
+            for (int i = 0; i < consumes.count; i++) {
+                sema_names.push_back(op->name + ".semaphore_" + std::to_string(i));
+                sema_vars.push_back(Variable::make(Handle(), sema_names.back()));
+            }
+
+            // debug(0) << "Body: " << body << "\n\n";
+
+            Stmt producer = GenerateProducerBody(op->name, sema_vars, cloned_acquires).mutate(body);
+            Stmt consumer = GenerateConsumerBody(op->name, sema_vars).mutate(body);
+
+            // debug(0) << "Producer: " << producer << "\n\n";
+            // debug(0) << "Consumer: " << consumer << "\n\n";
+
+            // Recurse on both sides
+            producer = mutate(producer);
+            consumer = mutate(consumer);
+
+            // Run them concurrently
+            body = Fork::make(producer, consumer);
+
+            for (const string &sema_name : sema_names) {
+                // Make a semaphore on the stack
+                Expr sema_space = Call::make(type_of<halide_semaphore_t *>(), "halide_make_semaphore",
+                                             {0}, Call::Extern);
+
+                // If there's a nested async producer, we may have
+                // recursively cloned this semaphore inside the mutation
+                // of the producer and consumer.
+                auto it = cloned_acquires.find(sema_name);
+                if (it != cloned_acquires.end()) {
+                    body = CloneAcquire(sema_name, it->second).mutate(body);
+                    body = LetStmt::make(it->second, sema_space, body);
+                }
+
+                body = LetStmt::make(sema_name, sema_space, body);
+            }
+
+            stmt = Realize::make(op->name, op->types, op->memory_type,
+                                 op->bounds, op->condition, body);
+        } else {
+            IRMutator::visit(op);
+        }
+    }
+
+public:
+    ForkAsyncProducers(const map<string, Function> &e) : env(e) {}
+};
+
+// Lowers semaphore initialization from a call to
+// "halide_make_semaphore" to an alloca followed by a call into the
+// runtime to initialize. TODO: what if something crashes before
```

Comment:
@abadams Do you mind looking over the new comment on InitializeSemaphores re: error handling in async. I added code to propagate errors across all branches of a fork. I'm not 100% sure it handles complex nested async/par_for structures however. (The situation is that some other part of the tree needs a semaphore that is not going to be released because another part failed, but the failure will not have poisoned the part of the tree that is waiting. I'm not convinced this can actually happen. Probably the thing to do is to take the async_parallel test and add errors into it.)

Reply:
I can't think through the possibilities in my head reliably for this sort of thing... Perhaps this feature needs a deadlock/error-handling fuzzer: construct random combinations of nested parallelism and check they can run, and check they clean up properly when errors are thrown internally. We would have to run it with a watchdog to avoid wedging buildbots though (SIGALRM?)